Adding model to segment bones

gattia opened the pull request:
Starting a pull request while working on this to get feedback. A few things are done differently than other models. 

- Currently loading model + weights from .h5 file (vs. building model & loading weights) 
    - This requires storing shape of original images used during training, reshaping inputs to that shape, segmenting, and then returning the shape. https://github.com/gattia/DOSMA/blob/adcebe3df0556eaddac95b3d7501d313f361a367/dosma/models/stanford_qdess_bone.py#L180-L215
    - Added option to "resample_images" which if True uses this strategy of using the original model (and resampling shape). If False, presumably we will re-build the original model and load the weights. This is essentially a placeholder for me to get code to re-generate the model(s) directly - I was worried at first I had something in the network that wasn't fully convolutional (Dense layer), but I think we are good.  Right now this throws an error (will implement in future). 
- Use argmax to create a single segmentation of each tissue (instead of thresholding each tissue separately). 
    -  https://github.com/gattia/DOSMA/blob/adcebe3df0556eaddac95b3d7501d313f361a367/dosma/models/stanford_qdess_bone.py#L199
- Added option `connected_only` to `model.generate_mask` this is because the bones produced a few spurious points in tests. This is on by default for all tissues. 
    - https://github.com/gattia/DOSMA/blob/adcebe3df0556eaddac95b3d7501d313f361a367/dosma/models/stanford_qdess_bone.py#L217
- The model segments medial/lateral tissues separately. 
    - To keep convention with other models, it then creates a "men" and "tc" tissue that just combines med/lat segments.
        - https://github.com/gattia/DOSMA/blob/adcebe3df0556eaddac95b3d7501d313f361a367/dosma/models/stanford_qdess_bone.py#L165-L176 
    - Also saved the "all" segmentation mask that is a single mask with all tissues (not one-hot/binary).  
        - https://github.com/gattia/DOSMA/blob/adcebe3df0556eaddac95b3d7501d313f361a367/dosma/models/stanford_qdess_bone.py#L152-L154
        
The main things I see myself wanting to do/add are: 

- [ ] Create function to build original model so we can skip resampling images up/down?
- [ ] Add a model using same architecture as other `stanford_qdess.py` models - I think the segmentation performance on this model vs. that are pretty much the same.

gattia committed:
Adding qdess bone segmentation model

gattia committed:
Adding qdess bone segmentation model

gattia committed:
Bone seg return all tissues
- qdess bone seg returns each individual tissue
- Added map of combined maps.

gattia committed:
Bone seg return all tissues
- qdess bone seg returns each individual tissue
- Added map of combined maps.

gattia committed:
Import bone model to models sub-module

gattia committed:
Import bone model to models sub-module

gattia committed:
Add bone seg model to models submodeuls namespace

gattia committed:
Add bone seg model to models submodeuls namespace

gattia committed:
Done super SegModel because not same funcs

gattia committed:
Done super SegModel because not same funcs

gattia committed:
Dont compile loaded bone seg model

gattia committed:
Dont compile loaded bone seg model

gattia committed:
fix tuple concatenation syntax

gattia committed:
fix tuple concatenation syntax

gattia committed:
expand dims for feature layer - use features first

gattia committed:
expand dims for feature layer - use features first

gattia committed:
Merge branch 'bone_seg' of https://github.com/gattia/DOSMA into bone_seg

gattia committed:
Add comments - update postprocess reshaping

gattia committed:
Merge remote-tracking branch 'refs/remotes/origin/bone_seg' into bone_seg

gattia committed:
Option (Default) get largest connected components

gattia committed:
Update docs, provide option for re-building model

gattia committed:
Updating documentation

gattia committed:
Apply linting

gattia committed:
autoformatting code with make

ad12 commented:
@gattia thanks for the PR! could you lint your files and provide some same screenshots of the segmentations

gattia committed:
Lint / autoformat

gattia committed:
Add bone segs figures

gattia commented:
Looks like it's getting hungup during CI because Python 3.6 isn't available anymore? 

Performed lint on my side + added a photo of the current segs. I'd love to add the two remaining things I have listed on my PR comments, but they're not necessary for this PR and maybe I can do at a later date.

gattia committed:
stanford_qdess_bone to return uint8 masks

gattia committed:
update StanfordQDessBoneINet2D to be flexible on model image size

gattia committed:
Merge branch 'bone_seg' of https://github.com/gattia/DOSMA into bone_seg

gattia committed:
refactor qdess bone seg model
- add ability to flood fill bone
- make post processing more flexible (for inherited classes)

gattia committed:
Add cube model

gattia committed:
Update tensorflow to be > 2.0.0 & remove import keras

gattia committed:
Update docs requirements to use same version of h5py ( not < 3)

gattia committed:
Black format updates

gattia committed:
update zip to use `strict=True`

gattia committed:
Fix flake8 linting - and add to ignore
Ignoring B907, which is to add !r to f-strings so they print literally.

gattia committed:
Add DS_Store to gitignore

gattia committed:
update make dev to install pytest

gattia committed:
Revert "update zip to use `strict=True`"

This reverts commit 150f44d7bbb8318f2decfa783896394aaefc6818.

gattia committed:
update np.int to int

gattia committed:
Add medvolume debugging print statements

gattia committed:
Get rid of strict=True for zip. so compatible with python<3.10

gattia committed:
Updating excel writer syntax to use .close vs .save

- if pd > 1.2 then use the new .close syntax. If older, then use old .save syntax.

https://stackoverflow.com/questions/76090979/xlsxwriter-object-has-no-attribute-save-did-you-mean-save

gattia committed:
Update h5py loading data syntax

gattia committed:
keep kers version logger - in case we go back to keras in future

gattia committed:
turn np index into tuple
- using non-tuple for multidimensional indexing was deprecated.
- also remove debug print statements

gattia committed:
update readme re: unit test environment variables.

gattia committed:
Updating np.bool to be bool b/c np.bool deprecated

gattia committed:
Fix deprecated scipy.ndimage.measurements
- if scipy >= 2.0.0 then import center_of_mass from scipy.ndimage
- else, if earlier then import from scipy.ndimage.measurements.

gattia committed:
Update install instructions to install from source.

gattia committed:
Update gitignore to not log nipype logs.

gattia committed:
ignore zip() strict=False in flake8 errors

- This is only available in python>3.10, therefore removing the
requirement in flake8

gattia committed:
Linting updates.

gattia committed:
Fixing sni import errors.

gattia committed:
Drop CI testing for python 3.6 and add 3.9/3.10

gattia committed:
Drop CI testing for python 3.6 and add 3.9/3.10

gattia committed:
Drop CI Documentation for 3.7
- Dont need docs for each specific version...

gattia committed:
Remove pinned numba & simpleitk from CI
- Number was due to an error in a specifc version (that numba is now past)
- SITK was due to change in basis (don't accept non-orthonormal) - but this hasnt been an issue running tests upto now.

gattia committed:
update `np.complex` to use builtin `complex`
- np.complex is deprecated.

gattia committed:
Flexibly get tensor shapes
- Different versions of tensorflow/keras were returning `tuple` or `tf.TensorShape`.
    - Create new function to handle both
    - refactor utils to create this function and the current model loading util functions.

gattia committed:
Update docs to specify we are not python 3.7+

gattia committed:
linting

gattia committed:
No codecov for forked repositroy pull request updates
- Codecoverage was throwing rate limiting errors and thus failing CI.
-https://github.com/ad12/DOSMA/actions/runs/10459944994/job/28965177349#step:7:83
- Therefore, add if statement to no push codecov under these circumstances.

gattia committed:
Update MedicalVolume.to_sitk

The current implementation worked somewhat to fix errors with specific image
acquisitions, but was somewhat specific to sagittal images. It has now been
updated to reformat and then export images to sitk as would be expected
for a specific image acquisition type (e.g., sagittal, coronal, axial).

Essentially, based on if its sag, ax, or coral, the `mv` will be reformatted.
Then, this reformatted image along with the affine/spacing information
a new sitk Image will be created. Finally, if the dicom was not collected in
this standard orientation, the user can still choose to flip the volume along
any of the three axes using boolean flags.

New tests were also written to ensure the reformatting is such that the
image is roughly what it would be like if SITK itself loaded it.

gattia committed:
Linting

gattia committed:
Update new sitk tests to use dicom data - thus not to run if data not available.

gattia committed:
Updating DESS T2 to not require Tg & GlArea.

This is based on a simplification of equations 6/7 of the original Sveinsson et al. paper "A Simple Analytic Method for Estimating T2 in the Knee from DESS". This is required because the new Mensa product sequence does not store these header fields.

gattia committed:
Updating DESS T2 to not require Tg & GlArea.

This is based on a simplification of equations 6/7 of the original Sveinsson et al. paper "A Simple Analytic Method for Estimating T2 in the Knee from DESS". This is required because the new Mensa product sequence does not store these header fields.

gattia committed:
Merge branch 'bone_seg' of https://github.com/gattia/DOSMA into bone_seg

gattia committed:
fix np round_ import error.

gattia committed:
Add plane-specific bone segmentation models for Stanford QDESS

- Introduced new subclasses for coronal, axial, and sagittal bone segmentation
- Added plane-specific TARGET_ORIENTATION and DEFAULT_IMAGE_SIZE
- Updated base class to use default image size when not provided
- Simplified model initialization with more flexible configuration

gattia committed:
Refactor bone segmentation volume processing

- fixed bug where reformat done to volume, but
- Improved volume processing logic for bone segmentation
- Simplified tissue combination and orientation handling
- Used logical OR for tissue combination instead of addition
- Reorganized volume creation and orientation reformatting steps

gattia committed:
Add framework for STAPLE-based multi-plane segmentation

- Introduced StanfordQDessBoneUNet2DSTAPLE class for combining masks from sagittal, coronal, and axial models
- Implemented mask generation using multiple plane-specific models
- Added tissue-specific mask combination logic using STAPLE approach
- Included memory management with garbage collection during model processing
- Prepared for future implementation of STAPLE algorithm for mask combination

gattia committed:
Implement STAPLE-based multi-plane bone segmentation with improved mask processing

- Replaced manual mask combination with SimpleITK's MultiLabelSTAPLE algorithm
- Added list of indices to exclude from STAPLE processing for each plane (model type).
- Simplified mask generation and combination logic
- Improved volume orientation handling and mask reformatting
- Removed hardcoded tissue combination dictionary

gattia committed:
Enhance STAPLE-based bone segmentation with configurable parameters and performance logging

- Added configurable tissue names and combination options to StanfordQDessBoneUNet2DSTAPLE
- Implemented optional verbose mode with timing information for mask generation steps
- Updated list of excluded indices for STAPLE processing
- Added time tracking for mask generation, processing, and STAPLE algorithm

gattia committed:
Update exclusion indices for STAPLE processing in StanfordQDessBoneUNet2DSTAPLE class

- Modified the list of indices to exclude for coronal and axial planes to improve mask processing accuracy.
- Ensured that the exclusion criteria align with the intended segmentation logic for better model performance.

- Sag: All labels
- Cor: All labels except patellar cartilage
- Axial: Only the bone labels

gattia committed:
Add verbose logging and extend StanfordQDessBoneUNet2DSTAPLE initialization

- Introduced optional verbose logging to track mask generation and processing times.
- Extended the constructor of StanfordQDessBoneUNet2DSTAPLE to accept tissue names and combinations for improved flexibility in mask generation.
- Updated the mask generation method to include timing information for performance monitoring.

gattia committed:
Update exclusion indices for axial processing in StanfordQDessBoneUNet2DSTAPLE class

Update so that the patellar cartilage is predicted by 2 networks (axial and sagittal).

gattia committed:
Enhance STAPLE processing in StanfordQDessBoneUNet2DSTAPLE class

- Cast masks to uint16 to prevent overflow during STAPLE algorithm execution.
- Set labels greater than 9 to 0 in the final mask to handle undecided labels.
- Convert final mask volume to uint8 for consistency in output.

gattia committed:
Refactor DICOM file reading to use pydicom.dcmread

- Replaced instances of pydicom.read_file with pydicom.dcmread for consistency and to leverage updated functionality.
- Updated corresponding test cases to reflect the change in the DICOM reading method.
